src/concurrency/ocf_metadata_concurrency.c

```diff
@@ -140,34 +151,69 @@ void ocf_metadata_concurrency_attached_deinit(
 void ocf_metadata_start_exclusive_access(
 		struct ocf_metadata_lock *metadata_lock)
 {
-        env_rwsem_down_write(&metadata_lock->global);
+	unsigned i;
+
+	for (i = 0; i < OCF_NUM_GLOBAL_META_LOCKS; i++) {
+		env_rwsem_down_write(&metadata_lock->global[i]);
+	}
 }
 
 int ocf_metadata_try_start_exclusive_access(
 		struct ocf_metadata_lock *metadata_lock)
 {
-	return env_rwsem_down_write_trylock(&metadata_lock->global);
+	unsigned i;
+	int error;
+
+	for (i = 0; i < OCF_NUM_GLOBAL_META_LOCKS; i++) {
+		error =  env_rwsem_down_write_trylock(&metadata_lock->global[i]);
+		if (error)
+			break;
+	}
+
+	if (error) {
+		while (i--) {
+			env_rwsem_up_write(&metadata_lock->global[i]);
+		}
+	}
+
+	return error;
 }
 
 void ocf_metadata_end_exclusive_access(
 		struct ocf_metadata_lock *metadata_lock)
 {
-        env_rwsem_up_write(&metadata_lock->global);
+	unsigned i;
+
+	for (i = OCF_NUM_GLOBAL_META_LOCKS; i > 0; i--)
+	        env_rwsem_up_write(&metadata_lock->global[i - 1]);
 }
 
-void ocf_metadata_start_shared_access(struct ocf_metadata_lock *metadata_lock)
+/* lock_idx determines which of underlying R/W locks is acquired for read. The goal
+   is to spread calls across all available underlying locks to reduce contention
+   on one single RW semaphor primitive. Technically any value is correct, but
+   picking wisely would allow for higher read througput:
+   * free running per-cpu counter sounds good,
+   * for rarely excercised code paths (e.g. management) any value would do.
+*/
+void ocf_metadata_start_shared_access(
+		struct ocf_metadata_lock *metadata_lock,
+		unsigned lock_idx)
 {
-        env_rwsem_down_read(&metadata_lock->global);
+        env_rwsem_down_read(&metadata_lock->global[lock_idx % OCF_NUM_GLOBAL_META_LOCKS]);
```

Comment:
Why doesn't `ocf_queue_get_next_shard_idx()` ensure that `lock_idx` is within range?

Reply:
done